File: WinForms/VB/GridMainDemo/KanbanModule/KanbanEditControl.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.Utils
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraGrid.Views.Grid

Namespace DevExpress.XtraGrid.Demos

    Public Partial Class KanbanEditControl
        Inherits DevExpress.XtraGrid.Views.Grid.EditFormUserControl

        Private OwnerView As DevExpress.XtraGrid.Views.Tile.TileView

        Private ReadOnly Property OwnerGrid As GridControl
            Get
                Return Me.OwnerView.GridControl
            End Get
        End Property

        Private Property Employees As DataTable

        Private Property Members As DataTable

        Private Property Checklist As DataTable

        '
        Public Sub New(ByVal ownerView As DevExpress.XtraGrid.Views.Tile.TileView, ByVal employees As System.Data.DataTable, ByVal members As System.Data.DataTable, ByVal checkList As System.Data.DataTable)
            Me.InitializeComponent()
            Me.OwnerView = ownerView
            Me.comboBoxEdit1.Properties.Items.AddRange(GetType(DevExpress.XtraGrid.Demos.TaskLabel).GetEnumValues())
            Me.Employees = employees
            Me.Members = members
            Me.Checklist = checkList
            Me.gridControl1.DataSource = Me.Checklist
            Me.memberTiles.AnimateArrival = False
            AddHandler Me.memberTiles.ItemClick, AddressOf Me.tileControl1_ItemClick
            AddHandler Me.VisibleChanged, Sub(s, e)
                If Me.Visible Then Me.ControlShown()
            End Sub
            AddHandler Me.gridChecklist.CellValueChanged, AddressOf Me.gridView1_CellValueChanged
            AddHandler Me.gridChecklist.RowCountChanged, Sub(s, e) Me.UpdateProgressBar()
        End Sub

        Private ReadOnly Property FocusedTaskId As Guid
            Get
                Return TryCast(Me.OwnerView.GetRow(CInt((Me.OwnerView.FocusedRowHandle))), DevExpress.XtraGrid.Demos.TaskRecord).Id
            End Get
        End Property

        Private Sub ControlShown()
            Me.layoutControl1.MenuManager = Me.OwnerGrid.MenuManager
            Me.PopulateMembers()
            Me.UpdateProgressBar()
        End Sub

        Private Sub PopulateMembers()
            Me.tileGroup.Items.Clear()
            Me.addMemberItem.Visible = True
            Me.tileGroup.Items.Add(Me.addMemberItem)
            Dim memberIDs As System.Collections.Generic.List(Of Integer) = Me.GetMemberIDs()
            For Each memberId As Integer In memberIDs
                Dim bytes = Me.Employees.Rows.Find(memberId)("Photo")
                Dim item = New DevExpress.XtraEditors.TileItem()
                item.Image = DevExpress.XtraEditors.Controls.ByteImageConverter.FromByteArray(TryCast(bytes, Byte()))
                item.ImageAlignment = DevExpress.XtraEditors.TileItemContentAlignment.MiddleCenter
                Dim superTip = New DevExpress.Utils.SuperToolTip()
                superTip.Items.AddTitle(Me.GetEmployeeFullName(memberId))
                item.SuperTip = superTip
                item.Tag = memberId
                Me.tileGroup.Items.Insert(0, item)
                Me.addMemberItem.Visible = If((Me.tileGroup.Items.Count = 5), False, True)
            Next
        End Sub

        Private Sub OnAddMemberMenuClick(ByVal sender As Object, ByVal e As System.EventArgs)
            Dim newId As Integer = CInt(TryCast(sender, DevExpress.Utils.Menu.DXMenuItem).Tag)
            Dim newRow As System.Data.DataRow = Me.Members.NewRow()
            newRow("TaskID") = Me.FocusedTaskId
            newRow("MemberID") = newId
            Me.Members.Rows.Add(newRow)
            Me.PopulateMembers()
        End Sub

        Private Sub addMemberItem_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.TileItemEventArgs)
            Dim memberIDs As System.Collections.Generic.List(Of Integer) = Me.GetMemberIDs()
            Dim menu As DevExpress.Utils.Menu.DXPopupMenu = New DevExpress.Utils.Menu.DXPopupMenu()
            menu.MenuViewType = DevExpress.Utils.Menu.MenuViewType.Menu
            For i As Integer = 0 To Me.Employees.Rows.Count - 1
                Dim row As System.Data.DataRow = Me.Employees.Rows(i)
                Dim id As Integer = row.Field(Of Integer)("EmployeeID")
                If memberIDs.Contains(id) Then Continue For
                Dim fullName As String = Me.GetEmployeeFullName(id)
                Dim memberMenuItem As DevExpress.Utils.Menu.DXMenuItem = New DevExpress.Utils.Menu.DXMenuItem(fullName, New System.EventHandler(AddressOf Me.OnAddMemberMenuClick)) With {.Tag = id}
                menu.Items.Add(memberMenuItem)
            Next

            Me.ShowPopup(menu)
        End Sub

        Private Sub tileControl1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.TileItemEventArgs)
            If e.Item Is Me.addMemberItem Then Return
            Dim menu As DevExpress.Utils.Menu.DXPopupMenu = New DevExpress.Utils.Menu.DXPopupMenu()
            menu.MenuViewType = DevExpress.Utils.Menu.MenuViewType.Menu
            Dim removeItem As DevExpress.Utils.Menu.DXMenuItem = New DevExpress.Utils.Menu.DXMenuItem("Remove from card", New System.EventHandler(AddressOf Me.OnRemoveItemClick)) With {.Tag = e.Item.Tag}
            menu.Items.Add(removeItem)
            Me.ShowPopup(menu)
        End Sub

        Private Sub OnRemoveItemClick(ByVal sender As Object, ByVal e As System.EventArgs)
            Dim memberId As Integer = CInt(TryCast(sender, DevExpress.Utils.Menu.DXMenuItem).Tag)
            Dim taskId As System.Guid = Me.FocusedTaskId
            Dim rowsToRemove As System.Collections.Generic.List(Of System.Data.DataRow) = Me.Members.AsEnumerable().Where(Function(row) row.Field(Of System.Guid)("TaskID") = taskId AndAlso row.Field(Of Integer)("MemberID") = memberId).ToList()
            For Each row As System.Data.DataRow In rowsToRemove
                Me.Members.Rows.Remove(row)
            Next

            Me.PopulateMembers()
        End Sub

        Private Sub ShowPopup(ByVal menu As DevExpress.Utils.Menu.DXPopupMenu)
            Dim parentControl As System.Windows.Forms.Control = Me.memberTiles
            Dim pt As System.Drawing.Point = parentControl.PointToClient(System.Windows.Forms.Control.MousePosition)
            CType(menu, DevExpress.Utils.Menu.IDXDropDownControl).Show(Me.OwnerGrid.MenuManager, parentControl, pt)
        End Sub

        Private Function GetEmployeeFullName(ByVal employeeId As Integer) As String
            Dim row = Me.Employees.Rows.Find(employeeId)
            Return String.Format("{0} {1}", row("FirstName"), row("LastName"))
        End Function

        Private Function GetMemberIDs() As List(Of Integer)
            Dim taskId As System.Guid = Me.FocusedTaskId
            Return Me.Members.AsEnumerable().Where(Function(row) row.Field(Of System.Guid)("TaskID") = Me.FocusedTaskId).[Select](Function(row) row.Field(Of Integer)("MemberID")).ToList()
        End Function

        Private Sub gridView1_CustomRowFilter(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.RowFilterEventArgs)
            Dim view As DevExpress.XtraGrid.Views.Grid.GridView = TryCast(sender, DevExpress.XtraGrid.Views.Grid.GridView)
            Dim dataView As System.Data.DataView = TryCast(view.DataSource, System.Data.DataView)
            Dim taskID As System.Guid = CType(dataView(e.ListSourceRow)("TaskID"), System.Guid)
            e.Visible =(taskID = Me.FocusedTaskId)
            e.Handled = True
        End Sub

        Private Sub gridView1_InitNewRow(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Grid.InitNewRowEventArgs)
            Me.gridChecklist.SetRowCellValue(e.RowHandle, Me.gridChecklist.Columns("TaskID"), Me.FocusedTaskId)
            Me.gridChecklist.SetRowCellValue(e.RowHandle, Me.gridChecklist.Columns("Caption"), "New item")
            Me.gridChecklist.SetRowCellValue(e.RowHandle, Me.gridChecklist.Columns("Checked"), False)
        End Sub

        Private Sub gridView1_CellValueChanged(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CellValueChangedEventArgs)
            Me.gridChecklist.UpdateCurrentRow()
            Me.UpdateProgressBar()
        End Sub

        Private Sub repositoryItemCheckEdit1_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs)
            Me.gridChecklist.PostEditor()
        End Sub

        Private Sub UpdateProgressBar()
            Dim taskId As System.Guid = Me.FocusedTaskId
            Dim filtered As System.Collections.Generic.List(Of System.Data.DataRow) = Me.Checklist.AsEnumerable().Where(Function(row) row.Field(Of System.Guid)("TaskID") = taskId).ToList()
            Dim totalCount As Integer = filtered.Count
            Dim checkedCount As Integer = System.Linq.Enumerable.ToList(Of System.Data.DataRow)(System.Linq.Enumerable.Where(Of System.Data.DataRow)(filtered, CType((Function(row) CBool((System.Data.DataRowExtensions.Field(Of Boolean)(row, CStr(("Checked"))) = True))), System.Func(Of System.Data.DataRow, System.[Boolean])))).Count
            Me.progressBarControl1.Properties.Maximum =(If(totalCount = 0, 1, totalCount))
            Me.progressBarControl1.Position = checkedCount
        End Sub
    End Class
End Namespace

File: WinForms/CS/GridMainDemo/KanbanModule/KanbanEditControl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using DevExpress.Utils;
using DevExpress.Utils.Menu;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.XtraGrid.Demos {
    public partial class KanbanEditControl : EditFormUserControl {
        DevExpress.XtraGrid.Views.Tile.TileView OwnerView;
        GridControl OwnerGrid {
            get { return OwnerView.GridControl; }
        }
        DataTable Employees { get; set; }
        DataTable Members { get; set; }
        DataTable Checklist { get; set; }
        //
        public KanbanEditControl(DevExpress.XtraGrid.Views.Tile.TileView ownerView,
            DataTable employees, DataTable members, DataTable checkList) {
            InitializeComponent();
            this.OwnerView = ownerView;
            this.comboBoxEdit1.Properties.Items.AddRange(typeof(TaskLabel).GetEnumValues());
            this.Employees = employees;
            this.Members = members;
            this.Checklist = checkList;
            this.gridControl1.DataSource = Checklist;
            this.memberTiles.AnimateArrival = false;
            this.memberTiles.ItemClick += tileControl1_ItemClick;
            this.VisibleChanged += (s, e) => { if(Visible) ControlShown(); };
            this.gridChecklist.CellValueChanged += gridView1_CellValueChanged;
            this.gridChecklist.RowCountChanged += (s, e) => { UpdateProgressBar(); };
        }
        Guid FocusedTaskId {
            get { return (OwnerView.GetRow(OwnerView.FocusedRowHandle) as TaskRecord).Id; }
        }
        void ControlShown() {
            this.layoutControl1.MenuManager = OwnerGrid.MenuManager;
            PopulateMembers();
            UpdateProgressBar();
        }
        void PopulateMembers() {
            tileGroup.Items.Clear();
            addMemberItem.Visible = true;
            tileGroup.Items.Add(addMemberItem);

            List<int> memberIDs = GetMemberIDs();
            foreach(int memberId in memberIDs) {
                var bytes = Employees.Rows.Find(memberId)["Photo"];
                var item = new TileItem();
                item.Image = ByteImageConverter.FromByteArray(bytes as byte[]);
                item.ImageAlignment = TileItemContentAlignment.MiddleCenter;
                var superTip = new SuperToolTip();
                superTip.Items.AddTitle(GetEmployeeFullName(memberId));
                item.SuperTip = superTip;
                item.Tag = memberId;
                tileGroup.Items.Insert(0, item);
                addMemberItem.Visible = (tileGroup.Items.Count == 5) ? false : true;
            }
        }
        void OnAddMemberMenuClick(object sender, EventArgs e) {
            int newId = (int)(sender as DXMenuItem).Tag;
            DataRow newRow = Members.NewRow();
            newRow["TaskID"] = FocusedTaskId;
            newRow["MemberID"] = newId;
            Members.Rows.Add(newRow);
            PopulateMembers();
        }
        void addMemberItem_ItemClick(object sender, TileItemEventArgs e) {
            List<int> memberIDs = GetMemberIDs();
            DXPopupMenu menu = new DXPopupMenu();
            menu.MenuViewType = MenuViewType.Menu;
            for(int i = 0; i < Employees.Rows.Count; i++) {
                DataRow row = Employees.Rows[i];
                int id = row.Field<int>("EmployeeID");
                if(memberIDs.Contains(id))
                    continue;
                string fullName = GetEmployeeFullName(id);
                DXMenuItem memberMenuItem = new DXMenuItem(fullName,
                    new EventHandler(OnAddMemberMenuClick)) { Tag = id };
                menu.Items.Add(memberMenuItem);
            }
            ShowPopup(menu);
        }
        void tileControl1_ItemClick(object sender, TileItemEventArgs e) {
            if(e.Item == addMemberItem)
                return;
            DXPopupMenu menu = new DXPopupMenu();
            menu.MenuViewType = MenuViewType.Menu;
            DXMenuItem removeItem = new DXMenuItem("Remove from card",
                new EventHandler(OnRemoveItemClick)) { Tag = e.Item.Tag };
            menu.Items.Add(removeItem);
            ShowPopup(menu);
        }
        void OnRemoveItemClick(object sender, EventArgs e) {
            int memberId = (int)(sender as DXMenuItem).Tag;
            Guid taskId = FocusedTaskId;
            List<DataRow> rowsToRemove = Members.AsEnumerable()
                .Where(row => row.Field<Guid>("TaskID") == taskId && row.Field<int>("MemberID") == memberId)
                .ToList();
            foreach(DataRow row in rowsToRemove)
                Members.Rows.Remove(row);
            PopulateMembers();
        }
        void ShowPopup(DXPopupMenu menu) {
            Control parentControl = memberTiles;
            Point pt = parentControl.PointToClient(Control.MousePosition);
            ((IDXDropDownControl)menu).Show(OwnerGrid.MenuManager, parentControl, pt);
        }
        string GetEmployeeFullName(int employeeId) {
            var row = Employees.Rows.Find(employeeId);
            return string.Format("{0} {1}", row["FirstName"], row["LastName"]);
        }
        List<int> GetMemberIDs() {
            Guid taskId = FocusedTaskId;
            return Members.AsEnumerable()
                .Where(row => row.Field<Guid>("TaskID") == FocusedTaskId)
                .Select(row => row.Field<int>("MemberID"))
                .ToList();
        }
        void gridView1_CustomRowFilter(object sender, Views.Base.RowFilterEventArgs e) {
            GridView view = sender as GridView;
            DataView dataView = view.DataSource as DataView;
            Guid taskID = (Guid)dataView[e.ListSourceRow]["TaskID"];
            e.Visible = (taskID == FocusedTaskId);
            e.Handled = true;
        }
        void gridView1_InitNewRow(object sender, InitNewRowEventArgs e) {
            gridChecklist.SetRowCellValue(e.RowHandle, gridChecklist.Columns["TaskID"], FocusedTaskId);
            gridChecklist.SetRowCellValue(e.RowHandle, gridChecklist.Columns["Caption"], "New item");
            gridChecklist.SetRowCellValue(e.RowHandle, gridChecklist.Columns["Checked"], false);
        }
        void gridView1_CellValueChanged(object sender, Views.Base.CellValueChangedEventArgs e) {
            gridChecklist.UpdateCurrentRow();
            UpdateProgressBar();
        }
        void repositoryItemCheckEdit1_CheckedChanged(object sender, EventArgs e) {
            gridChecklist.PostEditor();
        }
        void UpdateProgressBar() {
            Guid taskId = FocusedTaskId;
            List<DataRow> filtered = Checklist.AsEnumerable()
                .Where(row => row.Field<Guid>("TaskID") == taskId)
                .ToList();
            int totalCount = filtered.Count;
            int checkedCount = filtered.Where(row => row.Field<bool>("Checked") == true).ToList().Count;
            progressBarControl1.Properties.Maximum = (totalCount == 0 ? 1 : totalCount);
            progressBarControl1.Position = checkedCount;
        }
    }
}
